README.md
# express-test-helpers

[![Build Status](https://app.travis-ci.com/lykmapipo/express-test-helpers.svg?branch=master)](https://app.travis-ci.com/lykmapipo/express-test-helpers)
[![Dependencies Status](https://david-dm.org/lykmapipo/express-test-helpers.svg)](https://david-dm.org/lykmapipo/express-test-helpers)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/express-test-helpers/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/express-test-helpers?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/express-test-helpers)](https://github.com/lykmapipo/express-test-helpers/blob/master/LICENSE)

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/@lykmapipo/express-test-helpers)](https://www.npmjs.com/package/@lykmapipo/express-test-helpers)

Sensible test helpers for express

## Installation

```sh
$ npm install --save @lykmapipo/express-test-helpers
```

## Usage
```js
import { 
  app, testGet, testPost, testPatch, 
  testPut, testDelete, testOption, testHead 
} from '@lykmapipo/express-test-helpers';

app.get('/v1/users', (req, res)=> { res.ok(); });
app.post('/v1/users', (req, res)=> { res.created(); });

testGet('/v1/users')
    .expect('Content-Type', /json/)
    .expect(200)
    .end((err, res) => {
        if (err) throw err;
    });

testPost('/v1/users', { name: 'John Doe' })
    .expect('Content-Type', /json/)
    .expect(201)
    .end((err, res) => {
        if (err) throw err;
    });
```

- With Mocha
```js
import { clear, testGet } from '@lykmapipo/express-test-helpers';

describe('GET /v1/users', () => {
    
    beforeEach(() => clear());

    it('responds with json', done => {
        testGet('/v1/users')
            .expect('Content-Type', /json/)
            .expect(200, done);
    });
});
```

## Testing

- Clone this repository

- Install all development dependencies

```sh
$ npm install
```

- Then run test

```sh
$ npm test
```

## How to contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## LICENSE

MIT License

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 0.9.18 (2021-09-01)

#### 0.9.17 (2021-08-23)

##### Chores

* **deps:**  force latest version & audit fix ([7258dbd3](https://github.com/lykmapipo/express-test-helpers/commit/7258dbd3419fc43c7f809d717a38f778186bfc2e))
*  add status badge & improve configs ([5ee7513e](https://github.com/lykmapipo/express-test-helpers/commit/5ee7513ec8d8505266a5af7a8e188bb03de545e9))

#### 0.9.16 (2021-02-16)

#### 0.9.15 (2020-12-23)

##### Chores

* **deps:**
  *  force latest version & audit fix ([44dbe241](https://github.com/lykmapipo/express-test-helpers/commit/44dbe24194c4dda992e54ff05e4a58a879e66c58))
  *  bump ini from 1.3.5 to 1.3.7 ([82ae5742](https://github.com/lykmapipo/express-test-helpers/commit/82ae5742835ac5f4189a4cf314ed05f372275a45))

#### 0.9.14 (2020-09-13)

##### Chores

* **deps:**  force latest version & audit fix ([7773058e](https://github.com/lykmapipo/express-test-helpers/commit/7773058e4a4f063aa3a6d121354770ba2e7de828))

#### 0.9.13 (2020-09-03)

##### Chores

* **deps:**  force latest version & audit fix ([762f5b80](https://github.com/lykmapipo/express-test-helpers/commit/762f5b802c3894c4567f75981d27c302a44888cb))

#### 0.9.12 (2020-08-06)

##### Chores

* **deps:**  force latest version & audit fix ([240ace94](https://github.com/lykmapipo/express-test-helpers/commit/240ace94cd9650e68c049d310fe98742c76a8802))

#### 0.9.11 (2020-07-28)

##### Chores

* **deps:**  force latest version & audit fix ([567d6de8](https://github.com/lykmapipo/express-test-helpers/commit/567d6de84215cf716be009afef708369a134e27e))

#### 0.9.10 (2020-07-21)

##### Chores

* **deps:**  force latest version & audit fix ([564dc65f](https://github.com/lykmapipo/express-test-helpers/commit/564dc65f797aa51b70721363201a78e38747dbb4))

#### 0.9.9 (2020-07-12)

##### Chores

* **deps:**  force latest version & audit fix ([bb4208bd](https://github.com/lykmapipo/express-test-helpers/commit/bb4208bd8c7a0222c1a1812dd5f59de3ec9aa0b5))

#### 0.9.8 (2020-07-02)

##### Chores

* **deps:**  force latest version & audit fix ([aa4323b4](https://github.com/lykmapipo/express-test-helpers/commit/aa4323b449c9b983a8e8b25c59ddf13d1c8af71f))

#### 0.9.7 (2020-06-18)

##### Chores

* **deps:**  force latest version & audit fix ([b5417a03](https://github.com/lykmapipo/express-test-helpers/commit/b5417a034310d45a1f0aaa28c052820078e67408))

#### 0.9.6 (2020-06-09)

##### Chores

* **deps:**  force latest version & audit fix ([9517503f](https://github.com/lykmapipo/express-test-helpers/commit/9517503fb14d6c5472fcd2ed1c61d3a220e0e857))

#### 0.9.5 (2020-05-31)

##### Chores

* **deps:**  force latest version & audit fix ([338274d7](https://github.com/lykmapipo/express-test-helpers/commit/338274d7fabad07a4a1c699088e32ba3198403c1))

#### 0.9.4 (2020-05-19)

##### Chores

* **deps:**  force latest version & audit fix ([b2a10f93](https://github.com/lykmapipo/express-test-helpers/commit/b2a10f930481a82098cbaf5e5ea99dae7ff0a45c))

#### 0.9.3 (2020-05-10)

##### Chores

* **deps:**  force latest version & audit fix ([b0bbaa87](https://github.com/lykmapipo/express-test-helpers/commit/b0bbaa87901d949ea3d26a8ad39c0872d5348b43))

#### 0.9.2 (2020-04-24)

##### Chores

* **deps:**  force latest version & audit fix ([277bb1e9](https://github.com/lykmapipo/express-test-helpers/commit/277bb1e90658474e125a363278942818b4dc5f61))

#### 0.9.1 (2020-04-18)

##### Chores

* **deps:**  force latest version & audit fix ([eefa6297](https://github.com/lykmapipo/express-test-helpers/commit/eefa6297a32458100df7b7bafe65ba26784f4455))

#### 0.9.0 (2020-04-04)

##### Chores

* **deps:**
  *  force latest version & audit fix ([59ff1aaf](https://github.com/lykmapipo/express-test-helpers/commit/59ff1aaf37dfe83f29ae0169d19e68c30b6cf8de))
  *  force latest version & audit fix ([d1ad27c0](https://github.com/lykmapipo/express-test-helpers/commit/d1ad27c0dc8987c6abb81f3971e01ee4f79e1008))
  *  force latest version & audit fix ([de79aac5](https://github.com/lykmapipo/express-test-helpers/commit/de79aac5d32c58ea41b7672422ea64a7e9051d39))

##### Refactors

*  use latest uuid version ([a3323011](https://github.com/lykmapipo/express-test-helpers/commit/a33230114d1036a4a808326f1f0b4441ddf3ce77))

#### 0.8.21 (2020-03-30)

##### Chores

* **deps:**  force latest version & audit fix ([b5631b0e](https://github.com/lykmapipo/express-test-helpers/commit/b5631b0ef4ba4a1779252ba3fe5232226d1079d9))

#### 0.8.20 (2020-03-06)

##### Chores

* **deps:**  force latest version & audit fix ([277a04e0](https://github.com/lykmapipo/express-test-helpers/commit/277a04e08ab1d0427cdec73cf380a905ec767b30))

#### 0.8.19 (2020-02-28)

##### Chores

* **deps:**  force latest version & audit fix ([96ef2ae8](https://github.com/lykmapipo/express-test-helpers/commit/96ef2ae87d72b8aa554ba11d3c59914058005b03))

#### 0.8.18 (2020-02-25)

##### Chores

* **deps:**  force latest version & audit fix ([3876cf40](https://github.com/lykmapipo/express-test-helpers/commit/3876cf408d77f955682dc5931e2cd88fdacb3016))

##### New Features

*  expose mount router helper ([cd718615](https://github.com/lykmapipo/express-test-helpers/commit/cd718615bffaf240a6cfd0f439b548e06f85a66b))

#### 0.8.17 (2020-02-16)

##### Chores

* **deps:**  force latest version & audit fix ([92157ce5](https://github.com/lykmapipo/express-test-helpers/commit/92157ce540dd9823ca073b58da2234c4a3939a2a))

#### 0.8.16 (2020-02-04)

##### Chores

* **deps:**  force latest version & audit fix ([58e2acc7](https://github.com/lykmapipo/express-test-helpers/commit/58e2acc72a740a28bb7b050e559ecb07dc28bdaa))

#### 0.8.15 (2020-01-22)

##### Chores

* **deps:**  force latest version & audit fix ([7bc9ae81](https://github.com/lykmapipo/express-test-helpers/commit/7bc9ae81ef70eaed1c05f6c2bc8494e482b588ed))

#### 0.8.14 (2020-01-17)

##### Chores

* **deps:**  force latest version & audit fix ([b39bea72](https://github.com/lykmapipo/express-test-helpers/commit/b39bea72e4dec2595def752581db3645e19f0ced))

#### 0.8.13 (2020-01-12)

##### Chores


test/index.spec.js
import { createReadStream, readFileSync } from 'fs';
import multer from 'multer';
import { mount, Router } from '@lykmapipo/express-common';
import {
  expect,
  faker,
  app,
  clear,
  testOption,
  testHead,
  testGet,
  testPost,
  testPatch,
  testPut,
  testDelete,
  testUpload,
  testDownload,
  testStream,
} from '../src/index';

describe('generic test helpers', () => {
  beforeEach(() => clear());

  it('should test option request', (done) => {
    expect(testOption).to.exist.and.be.a('function');
    testOption('/v1/users').expect(204, done);
  });

  it('should test head request', (done) => {
    expect(testHead).to.exist.and.be.a('function');
    testHead('/v1/users').expect(404, done);
  });

  it('should test get request', (done) => {
    expect(testGet).to.exist.and.be.a('function');
    testGet('/v1/users').expect(404, done);
  });

  it('should test get request', (done) => {
    app.get('/v1/users', (req, res) => res.ok());
    testGet('/v1/users').expect(200, done);
  });

  it('should test post request', (done) => {
    expect(testPost).to.exist.and.be.a('function');
    testPost('/v1/users', { name: faker.name.findName() }).expect(404, done);
  });

  it('should test post request', (done) => {
    app.post('/v1/users', (req, res) => res.created());
    testPost('/v1/users', { name: faker.name.findName() }).expect(201, done);
  });

  it('should test patch request', (done) => {
    expect(testPatch).to.exist.and.be.a('function');
    testPatch('/v1/users/1', { name: faker.name.findName() }).expect(404, done);
  });

  it('should test patch request', (done) => {
    app.patch('/v1/users/:id', (req, res) => res.ok());
    testPatch('/v1/users/1', { name: faker.name.findName() }).expect(200, done);
  });

  it('should test put request', (done) => {
    expect(testPut).to.exist.and.be.a('function');
    testPut('/v1/users/1', { name: faker.name.findName() }).expect(404, done);
  });

  it('should test put request', (done) => {
    app.put('/v1/users/:id', (req, res) => res.ok());
    testPut('/v1/users/1', { name: faker.name.findName() }).expect(200, done);
  });

  it('should test delete request', (done) => {
    expect(testDelete).to.exist.and.be.a('function');
    testDelete('/v1/users/1').expect(404, done);
  });

  it('should test post request', (done) => {
    app.delete('/v1/users/:id', (req, res) => res.ok());
    testDelete('/v1/users/1').expect(200, done);
  });

  it('should test upload request', (done) => {
    const file = `${__dirname}/fixtures/test.txt`;
    const handleUpload = [
      multer({ dest: 'logs/' }).single('avatar'),
      (req, res) => res.ok({ ...req.body, ...req.file }),
    ];
    app.post('/v1/uploads', handleUpload);
    testUpload('/v1/uploads', { caption: 'avatar', attach: { avatar: file } })
      .expect(200)
      .end((error, { body }) => {
        expect(error).to.not.exist;
        expect(body.fieldname).to.be.eql('avatar');
        expect(body.originalname).to.be.eql('test.txt');
        expect(body.mimetype).to.be.eql('text/plain');
        expect(body.caption).to.be.eql('avatar');
        done(error, body);
      });
  });

  it('should test download request', (done) => {
    const file = `${__dirname}/fixtures/test.txt`;
    const fileContent = readFileSync(file).toString('base64');

    app.get('/v1/downloads', (req, res) => res.download(file));

    testDownload('/v1/downloads')
      .expect(200)
      .expect('Content-Type', 'text/plain; charset=UTF-8')
      .expect('Content-Disposition', 'attachment; filename="test.txt"')
      .end((error, { body }) => {
        expect(error).to.not.exist;
        expect(body.toString('base64')).to.be.equal(fileContent);
        done(error, body);
      });
  });

  it('should test stream download request', (done) => {
    const file = `${__dirname}/fixtures/test.txt`;
    const fileContent = readFileSync(file).toString('base64');

    app.get('/v1/streams', (req, res) => {
      res.attachment('test.txt');
      res.status(200);
      createReadStream(file).pipe(res);
    });

    testDownload('/v1/streams')
      .expect(200)
      .expect('Content-Type', 'text/plain; charset=utf-8')
      .expect('Content-Disposition', 'attachment; filename="test.txt"')
      .end((error, { body }) => {
        expect(error).to.not.exist;
        expect(body.toString('base64')).to.be.equal(fileContent);
        done(error, body);
      });
  });

  it('should test stream request', (done) => {
    const file = `${__dirname}/fixtures/test.txt`;
    const fileContent = readFileSync(file).toString('base64');

    app.get('/v1/streams', (req, res) => {
      res.type('test.txt');
      res.status(200);
      createReadStream(file).pipe(res);
    });

    testStream('/v1/streams')
      .expect(200)
      .expect('Content-Type', 'text/plain; charset=utf-8')
      .end((error, { body }) => {
        expect(error).to.not.exist;
        expect(body.toString('base64')).to.be.equal(fileContent);
        done(error, body);
      });
  });

  it('should clear attached routers', () => {
    const router = new Router({ version: '1.0.0' });
    mount(router);
    const beforeStack = app._router.stack;
    clear();
    const afterStack = app._router.stack;
    expect(beforeStack.length).to.not.be.equal(afterStack.length);
  });
});


test/middleware.spec.js
import { clear, testMiddleware } from '../src/index';

describe('middleware test helpers', () => {
  beforeEach(() => clear());

  const ipCheck = (req, res, next) => {
    res.set('X-IP-Checked', true);
    next();
  };

  it('should work on get', (done) => {
    const { testGet } = testMiddleware(ipCheck);
    testGet().expect('x-ip-checked', 'true').expect(200, done);
  });

  it('should work on post', (done) => {
    const { testPost } = testMiddleware(ipCheck);
    testPost().expect('x-ip-checked', 'true').expect(200, done);
  });

  it('should work on put', (done) => {
    const { testPut } = testMiddleware(ipCheck);
    testPut().expect('x-ip-checked', 'true').expect(200, done);
  });

  it('should work on patch', (done) => {
    const { testPatch } = testMiddleware(ipCheck);
    testPatch().expect('x-ip-checked', 'true').expect(200, done);
  });

  it('should work on delete', (done) => {
    const { testDelete } = testMiddleware(ipCheck);
    testDelete().expect('x-ip-checked', 'true').expect(200, done);
  });
});


lib/index.js
'use strict';

const lodash = require('lodash');
const uuid = require('uuid');
const testHelpers = require('@lykmapipo/test-helpers');
const expressCommon = require('@lykmapipo/express-common');
const supertest = require('supertest');
const pathToRegexp = require('path-to-regexp');

/**
 * @function clear
 * @name clear
 * @description Clear notFound, errorHandler and route handlers
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { clear } = require('@lykmapipo/express-test-helpers');
 *
 * beforeEach(() => clear());
 *
 */
const clear = () => {
  // eslint-disable-next-line no-underscore-dangle
  expressCommon.app._router.stack = lodash.filter(expressCommon.app._router.stack, (stack) => {
    const filtered =
      stack.name !== 'notFound' &&
      stack.name !== 'errorHandler' &&
      !stack.route &&
      !lodash.has(stack, 'handle.stack');
    return filtered;
  });
};

/**
 * @function testRequest
 * @name testRequest
 * @description Create generic test request with applied defaults
 * @return {Function} valid supertest request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testRequest } = require('@lykmapipo/express-test-helpers');
 *
 * const request = testRequest();
 * request.get('/users')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testRequest = () => {
  const request = supertest(expressCommon.testApp());
  return request;
};

/**
 * @function testOption
 * @name testOption
 * @description Create http option test request
 * @param {String} path valid path under test
 * @return {Function} valid supertest option request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testOption } = require('@lykmapipo/express-test-helpers');
 *
 * testOption('/v1/users')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 * testOption('/v1/users/1')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testOption = (path) => {
  const request = testRequest().options(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  return request;
};

/**
 * @function testHead
 * @name testHead
 * @description Create http option test request
 * @param {String} path valid path under test
 * @return {Function} valid supertest option request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testHead } = require('@lykmapipo/express-test-helpers');
 *
 * testHead('/v1/users')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 * testHead('/v1/users/1')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testHead = (path) => {
  const request = testRequest().head(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  return request;
};

/**
 * @function testGet
 * @name testGet
 * @description Create http get test request
 * @param {String} path valid path under test
 * @return {Function} valid supertest get request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testGet } = require('@lykmapipo/express-test-helpers');
 *
 * testGet('/v1/users')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 * testGet('/v1/users/1')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testGet = (path) => {
  const request = testRequest().get(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  return request;
};

/**
 * @function testPost
 * @name testPost
 * @description Create http get test request
 * @param {String} path valid path under test
 * @param {Object} [body={}] valid response body under test
 * @return {Function} valid supertest post request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testPost } = require('@lykmapipo/express-test-helpers');
 *
 * testPost('/v1/users', { name: 'John Doe', ... })
 *  .expect(201)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testPost = (path, body = {}) => {
  const request = testRequest().post(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  request.send(body);
  return request;
};

/**
 * @function testPatch
 * @name testPatch
 * @description Create http patch test request
 * @param {String} path valid path under test
 * @param {Object} [body={}] valid response body under test
 * @return {Function} valid supertest patch request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testPatch } = require('@lykmapipo/express-test-helpers');
 *
 * testPatch('/v1/users/1', { name: 'John Doe', ... })
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testPatch = (path, body = {}) => {
  const request = testRequest().patch(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  request.send(body);
  return request;
};

/**
 * @function testPut
 * @name testPut
 * @description Create http put test request
 * @param {String} path valid path under test
 * @param {Object} [body={}] valid response body under test
 * @return {Function} valid supertest put request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testPut } = require('@lykmapipo/express-test-helpers');
 *
 * testPut('/v1/users/1', { name: 'John Doe', ... })
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testPut = (path, body = {}) => {
  const request = testRequest().put(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  request.send(body);
  return request;
};

/**
 * @function testDelete
 * @name testDelete
 * @description Create http delete test request
 * @param {String} path valid path under test
 * @return {Function} valid supertest delete request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testDelete } = require('@lykmapipo/express-test-helpers');
 *
 * testDelete('/v1/users/1')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testDelete = (path) => {
  const request = testRequest().delete(path);
  request.set('Accept', 'application/json');
  request.set('Content-Type', 'application/json');
  return request;
};

/**
 * @function testMiddleware
 * @name testMiddleware
 * @description Create test requests for a middleware
 * @param {...Function} path valid express middleware
 * @return {Object} valid supertest requests
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testMiddleware } = require('@lykmapipo/express-test-helpers');
 *
 * const { testGet } = testMiddleware(ipFilter);
 *
 * testGet()
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testMiddleware = (...middlewares) => {
  const path = `/${uuid.v1()}`;
  const testReply = (req, res) => res.ok();
  expressCommon.app.all(path, ...[...middlewares, testReply]);
  return {
    testOption: () => testOption(path),
    testHead: () => testHead(path),
    testGet: () => testGet(path),
    testPost: (data = {}) => testPost(path, data),
    testPatch: (data = {}) => testPatch(path, data),
    testPut: (data = {}) => testPut(path, data),
    testDelete: () => testDelete(path),
    path,
  };
};

/**
 * @function testUpload
 * @name testUpload
 * @description Create http multiparty post(upload) test request
 * @param {String} path valid path under test
 * @param {Object} body valid multiparty body
 * @param {Object} body.attach valid attachments
 * @return {Function} valid supertest multiparty post request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.6.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testUpload } = require('@lykmapipo/express-test-helpers');
 *
 * const body =
 *  ({ caption: 'Avatar', attach: { avatar: { file: './avatar.png' } } });
 * testUpload('/v1/files', body)
 *  .expect(201)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testUpload = (path, body = {}) => {
  const request = testRequest().post(path);
  const { attach, ...fields } = body;

  // add fields
  lodash.forEach(fields, (value, key) => request.field(key, value));

  // add attachments
  lodash.forEach(attach, (value, key) => {
    // handle field name and options
    if (lodash.isPlainObject(value)) {
      const { file, ...fileOptions } = value;
      request.attach(key, file, fileOptions);
    }
    // handle field name and file path
    else {
      request.attach(key, value);
    }
  });

  // return request
  return request;
};

/**
 * @function testDownload
 * @name testDownload
 * @description Create http multiparty get(download) test request
 * @param {String} path valid path under test
 * @param {Object} optns valid options for test
 * @param {Function} [optns.encoding=binary] valid file encoding used for
 * parsing response body
 * @return {Function} valid supertest get request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.6.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testDownload } = require('@lykmapipo/express-test-helpers');
 *
 * testDownload('/v1/files')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testDownload = (path, optns = {}) => {
  // obtain options
  const { encoding = 'binary' } = optns;

  // prepare response body parser
  const parser = (response, cb) => {
    response.setEncoding(encoding);
    response.data = '';
    response.on('data', (chunk) => {
      response.data += chunk;
    });
    response.on('end', () => {
      cb(null, Buffer.from(response.data, encoding));
    });
  };

  // create request
  const request = testRequest().get(path).parse(parser);

  // return request
  return request;
};

/**
 * @function testStream
 * @name testStream
 * @description Create http multiparty stream test request
 * @param {String} path valid path under test
 * @param {Object} optns valid options for test
 * @param {Function} [optns.encoding=binary] valid file encoding used for
 * parsing response body
 * @return {Function} valid supertest get request
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.8.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const { testStream } = require('@lykmapipo/express-test-helpers');
 *
 * testStream('/v1/files')
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testStream = testDownload;

/**
 * @function testRouter
 * @name testRouter
 * @description Create test requests for express router
 * @param {Object|String} optns valid express router mount path or options
 * @param {Router} router valid express router
 * @return {Object} valid supertest requests
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.3.0
 * @static
 * @public
 * @example
 *
 * const { testRouter } = require('@lykmapipo/express-test-helpers');
 *
 * const { testGet } = testRouter('users', router);
 *
 * testGet()
 *  .expect(200)
 *  .end((err, res) => {
 *    if (err) throw err;
 *  });
 *
 */
const testRouter = (optns, router) => {
  // normalize options
  const options = lodash.isPlainObject(optns) ? optns : { resource: optns };

  // create paths
  let { pathSingle = `/${options.resource}/:id` } = options;
  let { pathList = `/${options.resource}` } = options;
  let { pathSchema = `/${options.resource}/schema` } = options;
  let { pathExport = `/${options.resource}/export` } = options;
  let { pathUpload = `/${options.resource}/upload` } = options;
  let { pathDownload = `/${options.resource}/download` } = options;
  let { pathStream = `/${options.resource}/stream` } = options;

  // handle versioned router
  pathSingle = pathToRegexp.compile(
    router.version ? `/${router.version}${pathSingle}` : pathSingle
  );
  pathList = pathToRegexp.compile(
    router.version ? `/${router.version}${pathList}` : pathList
  );
  pathSchema = pathToRegexp.compile(
    router.version ? `/${router.version}${pathSchema}` : pathSchema
  );
  pathExport = pathToRegexp.compile(
    router.version ? `/${router.version}${pathExport}` : pathExport
  );
  pathUpload = pathToRegexp.compile(
    router.version ? `/${router.version}${pathUpload}` : pathUpload
  );
  pathDownload = pathToRegexp.compile(
    router.version ? `/${router.version}${pathDownload}` : pathDownload
  );
  pathStream = pathToRegexp.compile(
    router.version ? `/${router.version}${pathStream}` : pathStream
  );

  // mout router for testing
  expressCommon.mount(router);

  // helpers
  const param = (val) => (lodash.isPlainObject(val) ? val : { id: val });
  const isSingle = (val) => {
    const params = param(val);
    const single = params && params.id;
    return single;
  };

  // pack test helpers
  return {
    testOption: (params = {}) => testOption(pathList(params)),
    testHead: (params = {}) => testHead(pathList(params)),
    testGet: (params = {}) =>
      isSingle(params)
        ? testGet(pathSingle(param(params)))
        : testGet(pathList(param(params))),
    testGetSchema: (params = {}) => testGet(pathSchema(param(params))),
    testGetExport: (params = {}) => testDownload(pathExport(param(params))),
    testExport: (params = {}) => testDownload(pathExport(param(params))),
    testDownload: (params = {}) => testDownload(pathDownload(param(params))),
    testStream: (params = {}) => testStream(pathStream(param(params))),
    testUpload: (data = {}) => testUpload(pathUpload(param(data)), data),
    testPost: (data = {}) => testPost(pathList(param(data)), data),
    testPatch: (id, data = {}) => testPatch(pathSingle(param(id)), data),
    testPut: (id, data = {}) => testPut(pathSingle(param(id)), data),
    testDelete: (id) => testDelete(pathSingle(param(id))),
  };
};

Object.defineProperty(exports, 'chai', {
  enumerable: true,
  get: function () {
    return testHelpers.chai;
  }
});
Object.defineProperty(exports, 'expect', {
  enumerable: true,
  get: function () {
    return testHelpers.expect;
  }
});
Object.defineProperty(exports, 'faker', {
  enumerable: true,
  get: function () {
    return testHelpers.faker;
  }
});
Object.defineProperty(exports, 'mock', {
  enumerable: true,
  get: function () {
    return testHelpers.mock;
  }
});
Object.defineProperty(exports, 'should', {
  enumerable: true,
  get: function () {
    return testHelpers.should;
  }
});
Object.defineProperty(exports, 'sinon', {
  enumerable: true,
  get: function () {
    return testHelpers.sinon;
  }
});
Object.defineProperty(exports, 'spy', {
  enumerable: true,
  get: function () {
    return testHelpers.spy;
  }
});
Object.defineProperty(exports, 'all', {
  enumerable: true,
  get: function () {
    return expressCommon.all;
  }
});
Object.defineProperty(exports, 'app', {
  enumerable: true,
  get: function () {
    return expressCommon.app;
  }
});
Object.defineProperty(exports, 'del', {
  enumerable: true,
  get: function () {
    return expressCommon.del;
  }
});
Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function () {
    return expressCommon.get;
  }
});
Object.defineProperty(exports, 'mount', {
  enumerable: true,
  get: function () {
    return expressCommon.mount;
  }
});
Object.defineProperty(exports, 'patch', {
  enumerable: true,
  get: function () {
    return expressCommon.patch;
  }
});
Object.defineProperty(exports, 'post', {
  enumerable: true,
  get: function () {
    return expressCommon.post;
  }
});
Object.defineProperty(exports, 'put', {
  enumerable: true,
  get: function () {
    return expressCommon.put;
  }
});
Object.defineProperty(exports, 'use', {
  enumerable: true,
  get: function () {
    return expressCommon.use;
  }
});
exports.clear = clear;
exports.testDelete = testDelete;
exports.testDownload = testDownload;
exports.testGet = testGet;
exports.testHead = testHead;
exports.testMiddleware = testMiddleware;
exports.testOption = testOption;
exports.testPatch = testPatch;
exports.testPost = testPost;
exports.testPut = testPut;
exports.testRequest = testRequest;
exports.testRouter = testRouter;
exports.testStream = testStream;
exports.testUpload = testUpload;
